### PR TITLE
tests: check files and dirs are cleaned for each test

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This is a new check to make sure all the files/dirs created are deleted
after the test is reset.

In a following PR the code could be moved to a test tool.

### CODE REVIEW

File: tests/lib/prepare-restore.sh [hunk 1]
```diff
@@ -634,6 +634,11 @@ prepare_suite_each() {
         "$TESTSTOOLS"/cleanup-state pre-invariant
     fi
     tests.invariant check
+
+    SNAP_MOUNT_DIR="$(os.paths snap-mount-dir)"
+    if os.query is-focal || os.query is.fedora; then
```
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh:639]: is-fedora?

File: tests/lib/prepare-restore.sh [hunk 2]
```diff
@@ -634,6 +634,11 @@ prepare_suite_each() {
...
+        inotifywait /root /var/lib/snapd /var/snap /home "$SNAP_MOUNT_DIR" -d -m -r -e CREATE -e DELETE --exclude "($PWD/task.yaml|$PWD/.*/.*.snap|$PROJECT_PATH/tests/lib/snaps/.*/.*.snap)" -o /tmp/fs.output
```
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh:640]: we may want to monitor /etc too, but that can be tricky when the test installs new packages

Comment (mardy) [inline tests/lib/prepare-restore.sh:640]: nitpick: better to first specify all the options, and keep the list of directories as the last parameters.

File: tests/lib/prepare-restore.sh [hunk 3]
```diff
@@ -675,6 +680,32 @@ restore_suite_each() {
         "$TESTSTOOLS"/cleanup-state pre-invariant
     fi
     tests.invariant check
+
+    if os.query is-focal || os.query is.fedora; then
+        set +x
+        pkill inotifywait
+        local created_files=$( sudo cat /tmp/fs.output | grep " CREATE " | awk '{ print $1$3 }' )
```
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh]: Could use a comment about the format of the logs, eg:
```
# the output file contains lines:
# /etc/ CREATE foo          # created file /etc/foo
# /etc/ CREATE,ISDIR foo   # created directory /etc/foo
```

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh:640]: Why track delete when we only check create events?

**Review by bboozzoo — COMMENTED** [review]
Nice idea!

File: tests/lib/prepare-restore.sh (hunk 1 above)
Comment (mardy) [inline tests/lib/prepare-restore.sh:639]: It's not immediately obvious why this should be restricted to focal and fedora. I'd suggest to create a has_inotify_tools function and use it here. Of course, for the time being the function body would check if the OS is focal or fedora, but in the future we can extend it as inotify-tools becomes available on more distros.

Comment (stolowski) [reply tests/lib/prepare-restore.sh:639]: Alternatively, check if `inotifywait` is available on PATH?

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:640]: Well, initially I tested differnet alternative, also tested files that could be deleted but at the end it was very complex to validate. I'll remove it. 
Thanks for the review

Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:640]: Nice, idea, thanks for the review

File: tests/lib/prepare-restore.sh (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:639]: Right, my idea was to add this to some systems and then make it available to any system. I'll add the dependencies and make it work on all the supported classic systems. 
Thanks for the review

File: tests/lib/prepare-restore.sh (hunk 3 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh]: Sure, thanks for the suggestion

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:640]: I'll add it, in case it causes trouble, I'll remove it.

File: tests/lib/prepare-restore.sh (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:639]: Nice catch, I added all the classic systems now.

File: tests/lib/prepare-restore.sh (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh]: I just run with /etc and the following paths are breaking tests:
File not deleted /etc/systemd/system/var-lib-snapd-snap-core-11378.mount
File not deleted /etc/systemd/system/multi-user.target.wants/var-lib-snapd-snap-core-11378.mount
File not deleted /etc/environment
File not deleted /etc/systemd/system/multi-user.target.wants/var-lib-snapd-snap-core-11378.mount
Dir not deleted /etc/systemd/system/snapd.service.d

Comment (bboozzoo) [reply tests/lib/prepare-restore.sh]: Looks like core 11378 could have been the revision the system came with and it got restored after the test finished?

> Dir not deleted /etc/systemd/system/snapd.service.d
isn't this something the test setup does?

File: tests/lib/prepare-restore.sh [hunk 4]
```diff
@@ -675,6 +687,42 @@ restore_suite_each() {
         "$TESTSTOOLS"/cleanup-state pre-invariant
     fi
     tests.invariant check
+
+    if os.query is-classic; then
```
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh:691]: Hm perhaps this should run earlier, before tests.cleanup restore and tests.backup restore, otherwise it may pick up file changes that were part of restore actually? Perhaps that's why it picked up the original core revision being manipulated.

File: tests/lib/tools/fs-state [hunk 5]
```diff
@@ -77,6 +80,66 @@ restore_file() {
...
+        retry -n 10 -wait 1 sh -c 'pgrep inotifywait &>/dev/null'  
+    fi
+}
+
+check_monitor() {
+    if [ -n "$(command -v inotifywait)" ]; then
+        # Stop inotifywait (it could be already stopped)
+        if pgrep inotifywait &>/dev/null; then
+            pkill inotifywait
+        fi
```
Comment (mardy) [inline tests/lib/tools/fs-state:104]: nitpick: I guess that with

    pkill inotifywait || true

you could get rid of the `if` block.

File: tests/lib/tools/fs-state [hunk 6]
```diff
@@ -77,6 +80,66 @@ restore_file() {
...
+        # the output file contains lines:
+        # /etc/ CREATE foo          # created file /etc/foo
+        # /etc/ CREATE,ISDIR foo    # created directory /etc/foo
+        created_files=$( sudo cat /tmp/fs.output | grep " CREATE " | awk '{ print $1$3 }' )
+        created_dirs=$( sudo cat /tmp/fs.output | grep " CREATE,ISDIR " | awk '{ print $1$3 }' )
+        missing=false
+
+        # Print the files that were created but not deleted
+        for file in $created_files; do
+            if [ -f "$file" ] || not grep -qx "$file" /tmp/fs.exclude; then
```
Comment (mardy) [inline tests/lib/tools/fs-state:116]: I'm a bit confused by this (probably because I don't know what is the expected behaviour of `exclude_monitor`): here it looks like you are considering it an error if a file which is not in the exclusion list has been created, regardless of whether it has been deleted or not. Shouldn't the error be emitted only if the created file has not been deleted? In other words, shouldn't the `||` be a `&&`?

File: tests/lib/tools/fs-state (hunk 5 above)
Comment (bboozzoo) [reply tests/lib/tools/fs-state:140]: That's a good idea, also maybe `pkill -9 inotifywait || true`

File: tests/lib/tools/fs-state [hunk 7]
```diff
@@ -77,6 +116,66 @@ restore_file() {
...
+        # In systems like arch-linux the inotifywait command fails when the output file does not exist
+        touch /tmp/fs.output
+        touch /tmp/fs.exclude
+
+        # Monitor files and directories creation
+        SNAP_MOUNT_DIR="$(os.paths snap-mount-dir)"
+
+        # $PWD is excluded because the backup dir is automatically restored
+        # .snap files created in tests/lib/snaps are excluded as they are not deleted to be reused
+        inotifywait -d -m -r -e CREATE -o /tmp/fs.output --exclude "($PWD|$PROJECT_PATH/tests/lib/snaps/.*/.*.snap)" /root /var/lib/snapd /var/snap /home "$SNAP_MOUNT_DIR"
```
Comment (bboozzoo) [inline tests/lib/tools/fs-state:130]: If we didn't need to support 14.04 this could be done via systemd-run, which also gives us a nice way to clean it up. TBH, maybe we don't want to use this feature on 14.04 after all?

File: tests/lib/tools/fs-state [hunk 8]
```diff
@@ -77,6 +116,66 @@ restore_file() {
...
+        # /etc/ CREATE foo          # created file /etc/foo
+        # /etc/ CREATE,ISDIR foo    # created directory /etc/foo
+        created_files=$( sudo cat /tmp/fs.output | grep " CREATE " | awk '{ print $1$3 }' )
+        created_dirs=$( sudo cat /tmp/fs.output | grep " CREATE,ISDIR " | awk '{ print $1$3 }' )
+        missing=false
+
+        # Print the files that were created but not deleted
+        for file in $created_files; do
+            if [ -f "$file" ] || not grep -qx "$file" /tmp/fs.exclude; then
+                echo "File not deleted $file"
```
Comment (bboozzoo) [inline tests/lib/tools/fs-state:153]:
```suggestion
                echo "File not cleaned up $file"
```

File: tests/lib/tools/fs-state [hunk 9]
```diff
@@ -77,6 +116,66 @@ restore_file() {
...
+        # the output file contains lines:
+        # /etc/ CREATE foo          # created file /etc/foo
+        # /etc/ CREATE,ISDIR foo    # created directory /etc/foo
+        created_files=$( sudo cat /tmp/fs.output | grep " CREATE " | awk '{ print $1$3 }' )
+        created_dirs=$( sudo cat /tmp/fs.output | grep " CREATE,ISDIR " | awk '{ print $1$3 }' )
+        missing=false
+
+        # Print the files that were created but not deleted
+        for file in $created_files; do
+            if [ -f "$file" ] || not grep -qx "$file" /tmp/fs.exclude; then
```
Comment (bboozzoo) [inline tests/lib/tools/fs-state:152]:
```suggestion
            if [ -f "$file" ] && not grep -qx "$file" /tmp/fs.exclude; then
```

File: tests/lib/tools/fs-state [hunk 10]
```diff
@@ -77,6 +116,66 @@ restore_file() {
...
+                echo "File not deleted $file"
+                missing=true
+            fi
+        done
+
+        # Print the directories that were created but not deleted
+        for dir in $created_dirs; do
+            if [ -d "$dir" ] || not grep -qx "$dir" /tmp/fs.exclude; then
```
Comment (bboozzoo) [inline tests/lib/tools/fs-state:160]:
```suggestion
            if [ -d "$dir" ] && not grep -qx "$dir" /tmp/fs.exclude; then
```

File: tests/lib/tools/fs-state [hunk 11]
```diff
@@ -1,25 +1,46 @@
 #!/bin/bash
 
 show_help() {
-    echo "usage: fs-state mock-dir [dirname]"
+    echo "usage: fs-state mock-dir <--new> [dirname]"
```
Comment (bboozzoo) [inline tests/lib/tools/fs-state:4]: Do we need --new in this PR or could it be a separate one?

Comment (sergiocazzolato) [reply tests/lib/tools/fs-state:4]: well, yes because there is a test wich fails in case we copy the file /dev/input/event1

File: tests/lib/pkgdb.sh [hunk 12]
```diff
@@ -494,6 +494,25 @@ distro_get_package_extension() {
     esac
 }
 
+distro_get_installed_packages() {
```
Comment (mardy) [inline tests/lib/pkgdb.sh:497]: nitpick: since this function is not returning the list of installed packages, I'd call it `distro_print_installed_packages()`

File: tests/lib/prepare-restore.sh [hunk 13]
```diff
@@ -605,6 +605,22 @@ prepare_suite() {
 prepare_suite_each() {
     local variant="$1"
 
+    # Create runtime files in case those don't exist
+    # This is for the first test of the suite, can't be added to the prepare_suite
+    # because some suites could not be preparing the suite such as the tools suite
```
Comment (mardy) [inline tests/lib/prepare-restore.sh:610]: assuming that I didn't misunderstood, I'd rephrase this to:

> This is for the first test of the suite. We cannot perform these operations in `prepare_suite` because not all suites are triggering it (for example the `tools` suite doesn't).

File: tests/lib/prepare-restore.sh [hunk 14]
```diff
@@ -654,6 +670,19 @@ restore_suite_each() {
...
+    if os.query is-classic; then
+        distro_get_installed_packages > installed-final.pkgs
+        diff -u installed-initial.pkgs installed-final.pkgs | grep -E "^\+" | tail -n+2 | cut -c 2- > installed-new.pkgs
+
+        # shellcheck disable=SC2002
+        packages="$(cat test | tr "\n" " ")"
+        if [ -n "$packages" ]; then
+            # shellcheck disable=SC2086
+            distro_purge_package $packages
+        fi
```
Comment (mardy) [inline tests/lib/prepare-restore.sh:683]: should we also remove the `installed-initial.pkgs` and `installed-final.pkgs` files, or is it not important?

File: tests/lib/prepare-restore.sh [hunk 15]
```diff
@@ -654,6 +670,19 @@ restore_suite_each() {
...
+    # Save all the installed packages and remove the new packages installed 
+    if os.query is-classic; then
+        distro_get_installed_packages > installed-final.pkgs
+        diff -u installed-initial.pkgs installed-final.pkgs | grep -E "^\+" | tail -n+2 | cut -c 2- > installed-new.pkgs
+
+        # shellcheck disable=SC2002
+        packages="$(cat test | tr "\n" " ")"
+        if [ -n "$packages" ]; then
+            # shellcheck disable=SC2086
+            distro_purge_package $packages
```
Comment (mardy) [inline tests/lib/prepare-restore.sh:682]: should we also mark the test as failed, if it did not properly remove all the packages it installed?

**Review by mardy — COMMENTED** [review]
Looks very god! Minor style suggestions, and a couple of questions :-)

File: tests/lib/tools/fs-state (hunk 10 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: this code changed

File: tests/lib/tools/fs-state (hunk 9 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: this code changed

File: tests/lib/tools/fs-state (hunk 11 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: this code has changed

File: tests/lib/tools/fs-state (hunk 8 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: this code has changed

File: tests/lib/tools/fs-state (hunk 7 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: this code has changed, we dont use inotifywait anymore, the new approach is using tree and diff

File: tests/lib/tools/fs-state (hunk 6 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: This approach has change, the new one is using tree and diff tools. Please take a look.

File: tests/lib/tools/fs-state (hunk 5 above)
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state]: inotifywait is not being used anymore

File: tests/lib/prepare-restore.sh (hunk 4 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh]: done

File: tests/lib/prepare-restore.sh (hunk 15 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:682]: well, in case it can't remove it should fail and the test will remain as failed

File: tests/lib/prepare-restore.sh (hunk 14 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:683]: thanks for reviewing, I prefer to keep it so in case of an error I can check the packages installed, at least at the beginning

File: tests/lib/pkgdb.sh (hunk 12 above)
Comment (sergiocazzolato) [reply tests/lib/pkgdb.sh:497]: sure, makes sense, thanks

File: tests/lib/prepare-restore.sh (hunk 13 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:610]: done

File: tests/lib/prepare-restore.sh [hunk 16]
```diff
@@ -654,6 +670,19 @@ restore_suite_each() {
     # restore test directory saved during prepare
     tests.backup restore
 
+    # Save all the installed packages and remove the new packages installed 
+    if os.query is-classic; then
+        distro_print_installed_packages > installed-final.pkgs
+        diff -u installed-initial.pkgs installed-final.pkgs | grep -E "^\+" | tail -n+2 | cut -c 2- > installed-new.pkgs
+
+        # shellcheck disable=SC2002
+        packages="$(cat test | tr "\n" " ")"
```
Comment (bboozzoo) [inline tests/lib/prepare-restore.sh:679]: the test file does not exist?

```suggestion
        packages="$(tr "\n" " " < installed-new.pkgs)"
```

File: tests/lib/reset.sh [hunk 17]
```diff
@@ -74,7 +74,12 @@ reset_classic() {
         systemctl start snap.mount.service
     fi
 
-    rm -rf /root/.snap/gnupg
+    # Clean root home
+    rm -rf /root/snap /root/.snap/gnupg /root/.{bash_history,local,cache,config}
+    # Clean test home
+    rm -rf /home/test/snap /home/test/.{bash_history,local,cache,config}
+
+    rm -rf /snap/*
```
Comment (bboozzoo) [inline tests/lib/reset.sh:82]: purge should have done it, perhaps there should be a `TODO` to inspect if there's anything left behind in the /snap directory

File: tests/smoke/install/task.yaml [hunk 18]
```diff
@@ -1,8 +1,8 @@
 summary: Check that installing and running a snap works
 
 restore: |
-    rm -f stderr.log
-    rm -f stdout.log
+    rm -f /home/test/stderr.log
```
Comment (bboozzoo) [inline tests/smoke/install/task.yaml:4]: I think files at both locations get created, so cleanup should happen in $PWD and in /home/test

File: tests/lib/prepare-restore.sh (hunk 16 above)
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:679]: nice catch, last migration I updated the names and forgot that one, thanks

File: tests/lib/reset.sh (hunk 17 above)
Comment (sergiocazzolato) [reply tests/lib/reset.sh:82]: Ok, I'll remove this nad check to see which files are being not deleted in /snap, then I can restore this and leave a comment with that info

File: tests/smoke/install/task.yaml (hunk 18 above)
Comment (sergiocazzolato) [reply tests/smoke/install/task.yaml:4]: The dir $PWD is automatically cleaned up during restore.

File: tests/lib/tools/boot-state [hunk 19]
```diff
@@ -107,7 +107,7 @@ bootenv_set() {
     if run_grub_editenv set "$var=$value"; then
         return
     elif [ -s "$grubenv_file" ]; then
-        sed -i "/^$var=/d" "$grubenv_file"
+        sed --follow-symlinks -i "/^$var=/d" "$grubenv_file"
```
Comment (stolowski) [inline tests/lib/tools/boot-state:110]: Is this change related to this PR?

File: tests/lib/prepare-restore.sh [hunk 20]
```diff
@@ -605,6 +605,22 @@ prepare_suite() {
 prepare_suite_each() {
     local variant="$1"
 
+    # Create runtime files in case those don't exist
+    # This is for the first test of the suite. We cannot perform these operations in prepare_suite
+    # because not all suites are triggering it (for example the tools suite doesn't).
+    if [ ! -f "$RUNTIME_STATE_PATH/runs" ] || [ ! -f "$RUNTIME_STATE_PATH/journalctl_cursor" ]; then
+        touch "$RUNTIME_STATE_PATH/runs"
+        touch "$RUNTIME_STATE_PATH/journalctl_cursor"
+    fi
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/prepare-restore.sh:614]: no, it is not important, I'll remove the if, thanks for the comment

File: tests/lib/tools/boot-state (hunk 19 above)
Comment (sergiocazzolato) [reply tests/lib/tools/boot-state:110]: yes, it happened that otherwise it was breaking the symlinks. After sed (removing a line), the file was not a symlink anymore, it was a regular file.

File: tests/lib/tools/boot-state [hunk 21]
```diff
@@ -132,7 +132,7 @@ bootenv_unset() {
     if run_grub_editenv "$grubenv_file" unset "$var"; then
         return
     elif [ -s "$grubenv_file" ]; then
-        sed -i "/^$var=/d" "$grubenv_file"
+        sed --follow-symlinks -i "/^$var=/d" "$grubenv_file"
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/boot-state:135]: same answer

File: tests/lib/tools/fs-state [hunk 22]
```diff
@@ -77,6 +81,60 @@ restore_file() {
...
+    local dir=$1
+
+    rm -rf "$dir"
+    mkdir "$dir"
+    # .snap files created in tests/lib/snaps are excluded as they are not deleted to be reused
+    # runtime-state dir contains cursor and run log which are created during the first test
+    tree --noreport -if /etc -I "*~" > "$dir"/etc
+    tree --noreport -if /boot -I "*~" > "$dir"/boot
+    tree --noreport -if /home -I "*~" | grep -v -e "$PROJECT_PATH/tests/lib/snaps/.*/.*.snap" -e "$PWD/.*.snap" > "$dir"/home
+    tree --noreport -if /root -I "*~" > "$dir"/root
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/fs-state:94]: well, yes, I started with tree because I was printing tree structure but then after face some errors I added -i and it became a find. I'll update this.
